src/librustc_mir/interpret/operand.rs

```diff
@@ -445,7 +445,14 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         layout: Option<TyAndLayout<'tcx>>,
     ) -> InterpResult<'tcx, OpTy<'tcx, M::PointerTag>> {
         let base_op = match place.local {
-            mir::RETURN_PLACE => throw_ub!(ReadFromReturnPlace),
+            mir::RETURN_PLACE => {
+                if let Some(place) = self.frame().return_place {
+                    self.place_to_op(place)?
+                } else {
+                    // Tried to access return place, but the function cannot return.
+                    throw_ub!(Unreachable)
```

Comment:
"Unreachable" is a bit non-descript as an error here, I'd rather just use that for actual unreachable code and not for this which is more like a dangling memory access.

It would be consistent with `eval_place` to return `place_to_op(Place::null(&*self))` here, which makes sure any actual read causes a dangling-pointer error. Maybe we should just make `return_place` not an `Option` any more and use a NULL place as default?

Reply:
In fact we can almost get rid of that `null` hack, except that call_once shim does this:
```rust
[INFO  rustc_mir::interpret::eval_context] ENTERING(10) <[closure@tests/run-pass/issue-miri-1075.rs:2:24: 2:48] as std::ops::FnOnce<()>>::call_once - shim
[INFO  rustc_mir::interpret::step] // executing bb0
[INFO  rustc_mir::interpret::step] _3 = &mut _1
[INFO  rustc_mir::interpret::step] _0 = const <Self as std::ops::FnMut<Args>>::call_mut(move _3, move _2) -> [return: bb1, unwind: bb3]
```
@eddyb would it be reasonable to ask that MIR of a function without a return place, must never actually evaluate that return place, not even to call another function without a return place? It is unclear, above, what the `_0` place should evaluate to given that there is no return place for this function.

Reply:
This is because it's returning an uninhabited type? You need to manufacture some scratch space.

Because of partial initialization, it's possible for the function to write fields of `_0` before diverging, and those fields should be in real memory.

I think codegen also creates an `alloca` when it doesn't have a return place to write to.

Reply:
Wow, what a mess. It seems unreasonable to put this burden on the Abstract Machine, IMO. When a `Call` happens without a `destination`, it should be UB for the callee to try to access that destination.

Of course, if we always treated the return place like normal arguments, allocating it ourselves and copying the result after the function returns, then this would work out with no special casing at all. All of this ugliness is only needed because we are trying to avoid the copy *in the specification*.

Reply:
> When a `Call` happens without a destination, it should be UB for the callee to try to access that destination.

The lack of a destination just means the function doesn't return.
The *MIR* body of a function may not know it doesn't return, because generics.

Maybe we should always have a destination `Place` but not a `BasicBlock` target?

Reply:
> Maybe we should always have a destination Place but not a BasicBlock target?

If diverging functions may indeed access their return place as you say, then yes -- I think that would be most consistent.